### DISCUSSION TITLE
Setting creation date of image from file name upon import?

### DISCUSSION THREAD

**Discussion by AlexisMaizel** [RhetTbull/osxphotos#840] [Q&A]
Hi! 
I have a bunch of images lacking metadata but the file name contains the image capture date:
`001_2009-03-29_someText.png`
I'd like to set the {created} metadata from the file name. 

What would be the best strategy? 
1)  Upon import with a custom template function? Is that possible? 
2)  Once imported in Photos,   parsing all images and set  the creation date?

Second part of my question, experimenting with (1), I wanted to see if I understand the template function by setting  --album: 
` osxphotos import 0-ToPhotos/* --walk --album "{filepath|function:/Users/maizel/Code/scripts/template_function.py::get_shot_date}" --check-templates` 

where `template_function.py` is
```
import pathlib
import re

def get_shot_date(filepath: pathlib.Path):
    filename = filepath.name
    regex = r"^\d{3}_(\d{4}\-\d{2}\-\d{2})_.*\.png"
    match = re.search(regex, filename)
    if match is not None:
        date = match.group(1)
    else:
        date=""
    return date
```
 Running this returns an error: 
```
/Users/maizel/Code/scripts/template_function.py:20 in get_shot_date                              │
│                                                                                                  │
│   17 │   Returns:                                                                                │
│   18 │   │   str or list of str of values that should be substituted for the {function} templ    │
│   19 │   """                                                                                     │
│ ❱ 20 │   filename = filepath.name                                                                │
│   21 │   regex = r"^\d{3}_(\d{4}\-\d{2}\-\d{2})_.*\.png"                                         │
│   22 │   match = re.search(regex, filename)                                                      │
│   23 │   if match is not None:                                                                   │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'list' object has no attribute 'name'
```
I do not understand, as I'm passing `{filepath}` a Pathlib object, correct? 
What am I missing?

**Comment by RhetTbull** [comment]
There are two types of template functions: filters and templates. You're using a template filter which gets passed the filepath.  The signature for a filter is `def myfilter(values: List[str]) -> List[str]:`  In other words, it receives a list of strings and outputs a list of strings. The reason for this is that the template system supports multi-valued templates and filters. For example, `{keyword}` expands to a list of keywords.  So all values are treated as lists until the template is finally rendered.  If there is only a single value after rendering, then there's only one album but it would be possible to have multiple albums depending on how the template was rendered.

So your function needs to get a list of strings (which in this case is only one item, the str value of filepath) and output a list of strings. So something like this:

```python
import typing as t
import pathlib

def get_shot_date(filepath: t.List[str]) -> t.List[str]:
    filename = pathlib.Path(filepath[0]).name
    regex = r"^\d{3}_(\d{4}\-\d{2}\-\d{2})_.*\.png"
    match = re.search(regex, filename)
    if match is not None:
        date = match.group(1)
    else:
        date=""
    return [date]
```

To answer the question of setting dates on imported images, the `import` command doesn't currently have a post-processing function but that's a great idea and I'll open an issue. In other words, after the photo is imported, your custom function is called with a Photo object representing the newly imported photo and you can edit the date, etc.  It might also be worthwhile to have an option that lets you set the date with a template.

But osxphotos does have a `timewarp` function that can adjust date and time and timezone and can call a custom function to do this.  See [timewarp help](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-timewarp)

**Reply by oPromessa** [reply]
@AlexisMaizel 

What I normally do is to set the EXIF create, original and modified dates with exiftool prior to loading them onto Photos. This way the original file will always have the correct dates. 

See: https://exiftool.org/forum/index.php?topic=8456.0;all

And also https://exiftool.org/faq.html#Q5

I've created some bash functions to do this depending on the file names. 

And after that, when I have several files for which I only know the day and approximate time I normally set the time with a few seconds apart with the command

```
exiftool -overwrite_original_in_place '-datetimeoriginal+<0:0:${filesequence;$_*=30}' -v files*.jpg
```

**Comment by RhetTbull** [comment]
@AlexisMaizel @oPromessa  In [v0.54.4](https://github.com/RhetTbull/osxphotos/releases/tag/v0.54.4) I've added a `--parse-date` option that can be used to set photo dates from the filename.  For example, in your example above, `001_2009-03-29_someText.png` could be parsed as `osxphotos import *.png --parse-date "_%Y-%m-%d_"`  You can chain multiple patterns together (to be tried in sequence using `|`.  For example, if you also had files in format `001_20090329.png`, you could use `osxphotos import *.png --parse-date "_%Y-%m-%d_|_%Y%m%d."`  The pattern is a superset of [strptime/strftime](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes) format codes.  See `osxphotos help import` for additional details.  I had to write a [custom date parser](https://github.com/RhetTbull/strpdatetime) to make this work which turned out to be  fun exercise.  

I've also added a new `--post-function` that will allow you to run a custom python function after import. The function will be passed a `Photo` object which allows you to set the date (and many other properties) on the newly imported photo.  See example [here](https://github.com/RhetTbull/osxphotos/blob/master/examples/post_function_import.py).

**Reply by AlexisMaizel** [reply]
This is fantastic! Thanks a lot. This will simplify my workflow.
